File: small_problems/easy7/how_many.rb
def count_occurrences(arr)
  hash = {}

  arr.each do |vehicle|
    if hash.key?(vehicle)
      hash[vehicle] += 1
    else
      hash[vehicle] = 1
    end
  end

  hash.each { |key, value| puts "#{key} => #{value}" }
end

vehicles = [
  'car', 'car', 'truck', 'car', 'SUV', 'truck',
  'motorcycle', 'motorcycle', 'car', 'truck'
]

# alternate

def count_occurrences(array)
  occurrences = {}

  array.uniq.each do |element|
    occurrences[element] = array.count(element)
  end

  occurrences.each do |element, count|
    puts "#{element} => #{count}"
  end
end

# further exploration

def count_occurrences(arr)
  hash = {}

  arr.each do |vehicle|
    vehicle = vehicle.downcase
    
    if hash.key?(vehicle)
      hash[vehicle] += 1
    else
      hash[vehicle] = 1
    end
  end

  hash.each { |key, value| puts "#{key} => #{value}" }
end

count_occurrences(vehicles)

# expected output:
# car => 4
# truck => 3
# SUV => 1
# motorcycle => 2

File: javascript_small_problems/easy4/how_many.js
const vehicles = ['car', 'car', 'truck', 'car', 'SUV', 'truck',
                'motorcycle', 'suv', 'motorcycle', 'car', 'truck'];

function countOccurrences(array) {
  let result = {};
  for (let value of array) {
    result[value] = result[value] || 0;
    result[value] += 1;
  }

  for (let key in result) {
    console.log(`${key} => ${result[key]}`);
  }
}

countOccurrences(vehicles);

// console output
// car => 4
// truck => 3
// SUV => 1
// motorcycle => 2
// suv => 1